### PR TITLE
mmap elimination

### PR DESCRIPTION
**Pull request by abudnik**
1) Problem description
If error happens at block i/o device level, then access to memory mapped file, which is stored in this device, may lead to program termination with SIGBUS signal.
When elliptics uses multiple backends, then problems of some backend may lead to malfunction of all backends. Elliptics (dnet_ioserv) nodes used to work with one backend only (some time ago), therefore increasing network traffic (route list updates), but isolated problems of a particular backend.

2) Problem solution
This PR provides eblob version without usage of memory-mapped files. Now elliptics/eblob use syscalls for read/write operations (pread, pwrite), which returns error code in case of any i/o errors.

3) Estimation of i/o operation cost (pread vs. mmap)
Using pread/pwrite for i/o operations leads to additional overhead and resources consumption comparing to using of memory-mapped files. Every i/o operation consists of following steps:
system call, data copying between user space buffers and kernel space, memory allocacation and dealocation for data buffers.

4) Affected operations after elimination of mmap
* Index element lookup in index file => all operation of read and write to eblob backend
* Sorting of index files
* Record iteration
* Checksumming

5) Performance comparison on synthetic load tests (pread vs. mmap)
We compared performance of pread vs. mmap version of eblob on following load tests:
* read at 10k rps: records 0.5-2.5 Mb
* read at 12k rps: records 100 bytes - 10 Kb
* write from 4 writing nodes (elliptics client): records 100-500 bytes
* write from 1 writing node: records 100-500 bytes
* iterate all keys

All read and write load tests where aimed at single elliptics node with 16 backends.
Write tests lasts short time (few minutes) and long time (30 minutes) at high rps.

Those tests (accessing small keys) has shown some performance divergence between mmap and pread versions.
Tests with heavy-weight keys didn't lead to any performances divergence.
Analyzing performance tests leads to next conclusion: in some tests response times increased a little bit, in other tests response times even decreased a little bit; if average and median rps differs, then it differs at a gap of 4-5 % in both sides (faster or slower).
Also pread version did not lead to changes in system load (cpu usage, context switches, interrupts, memory consumption and iowaits).

### CODE REVIEW

File: bindings/python/eblob_python.cpp [hunk 1]
```diff
@@ -52,6 +53,29 @@ static void eblob_extract_id(const struct eblob_id &e, struct eblob_key &id)
 	eblob_extract_arr(e.id, id.id, &len);
 }
 
+/**
+ * __eblob_read_ll() - interruption-safe wrapper for pread(2)
+ */
+static int __eblob_read_ll(int fd, char *data, size_t size, off_t offset)
```
Comment (bioothod) [inline bindings/python/eblob_python.cpp]: How will it behave when I press Ctrl-C while python binding reads data from fd?

Comment (bioothod) [inline bindings/python/eblob_python.cpp]: Why did you copy-paste this whole function from one already present in eblob code?
Difference between 'blocking' or ignoring signals in client and server code is fundamental.

File: bindings/python/eblob_python.cpp [hunk 2]
```diff
@@ -72,13 +96,18 @@ struct eblob_py_iterator : eblob_iterate_control, boost::python::wrapper<eblob_i
 	}
 
 	static int iterator(struct eblob_disk_control *dc, struct eblob_ram_control *rc __attribute__((unused)),
-			void *data, void *priv, void *thread_priv __attribute__((unused)))
+			    int fd, uint64_t data_offset, void *priv, void *thread_priv __attribute__((unused)))
 	{
+		std::auto_ptr<char> p(new char[dc->data_size]);
```
Comment (Kobolog) [inline bindings/python/eblob_python.cpp]: Note that using `std::auto_ptr<T>` is considered a bad practice for quite some time.

File: library/index.c [hunk 3]
```diff
@@ -380,25 +397,53 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
+			if (num > b->cfg.index_block_size)
+				num = b->cfg.index_block_size;
+
+			hdr_block_size = num * hdr_size;
+			sorted = hdr_block;
+			end = sorted + (num - 1);
+
+			bytes = pread(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
+			if (bytes != hdr_block_size) {
+				if (err < 0)
```
Comment (bioothod) [inline library/index.c]: 'err' has not yet correctly initialized here

File: library/index.c [hunk 4]
```diff
@@ -323,24 +323,25 @@ int eblob_index_blocks_fill(struct eblob_base_ctl *bctl)
 }
 
 
-static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
-		struct eblob_base_ctl *bctl, struct eblob_disk_control *dc,
+static int eblob_find_on_disk(struct eblob_backend *b,
```
Comment (bioothod) [inline library/index.c]: Please describe how does this function work while not binary searching over mapped area.

File: library/index.c [hunk 5]
```diff
@@ -412,10 +457,40 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
+			hdr_block_size = hdr_block_offset;
+			num = hdr_block_size / hdr_size;
+			hdr_block_offset = 0;
+		} else {
+			hdr_block_offset -= hdr_block_size;
+		}
+
+		bytes = pread(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
+		if (bytes != hdr_block_size) {
+		    if (err < 0)
```
Comment (bioothod) [inline library/index.c]: The same bug with 'err' variable

File: library/index.c [hunk 6]
```diff
@@ -645,19 +719,33 @@ int eblob_generate_sorted_index(struct eblob_backend *b, struct eblob_base_ctl *
...
 		}
 
+		err = __eblob_write_ll(fd, sorted_index, index_size, 0);
+		if (err) {
+			EBLOB_WARNC(b->cfg.log, EBLOB_LOG_ERROR, -err, "defrag: indexsort: write after binlog apply: index: %d, size: %llu: %s",
+					bctl->index, (unsigned long long)index_size, file);
+			goto err_unlock_hash;
+		}
+
+		if ((err = fsync(fd)) == -1) {
```
Comment (bioothod) [inline library/index.c]: Split this line to multiple statements

File: tests/stress/stress.c [hunk 7]
```diff
@@ -584,11 +583,19 @@ static int iterate_callback(struct eblob_disk_control *dc,
...
+				if (error == -1) {
+					errx(EX_SOFTWARE, "pread has been failed for: %s (%s), flags: %s, error: %d",
+					     item->item->key, eblob_dump_id(item->item->ekey.id), item->item->hflags, errno);
+				}
 				error = memcmp(data, item->item->value, item->item->size);
 				if (error != 0) {
 					errx(EX_SOFTWARE, "data verification has been failed for: %s (%s), flags: %s",
 					    item->item->key, eblob_dump_id(item->item->ekey.id), item->item->hflags);
 				}
+				free(data);
```
Comment (bioothod) [inline tests/stress/stress.c]: Did you run eblob_stress with this whole new code?
Were there any errors? What were the options?

File: library/index.c (hunk 4 above)
Comment (abudnik) [inline library/index.c]: I have found 2 issues in eblob_find_on_disk: one problem exists in current code and one thing I forget to add into this PR.
First one is about this lines:
if (num > b->cfg.index_block_size)
num = b->cfg.index_block_size;
// then bsearch only in this block, which contains only small part of index file.

Second one that I've forgotten to add implementation of bsearch with pread on each binary search iteration. So it reads whole index file into memory every time in this PR.
Also binary searching causes random read pattern, therefore index file must be prefetched into page cache before bin search.

File: tests/stress/stress.c (hunk 7 above)
Comment (abudnik) [inline tests/stress/stress.c]: I used those commands copied from tests/ci/travis.sh:


$(find . -name eblob_stress) -m0 -f1000 -D0 -I300000 -o20000 -i1000 -l0 -r 1000 -S10 -F87
$(find . -name eblob_stress) -m0 -f100 -D0 -I30000 -o2000 -i100 -l0 -r 100 -S100 -F14
$(find . -name eblob_stress) -m0 -f1000 -D0 -I300000 -o20000 -i1000 -l0 -r 1000 -S10 -F2135
$(find . -name eblob_stress) -m0 -f100 -D0 -I30000 -o2000 -i100 -l0 -r 100 -S100 -F2062
$(find . -name eblob_stress) -f1000 -D0 -I100000 -i64 -r 40 -S10 -F64 -T32 -l0 -o 0
$(find . -name eblob_stress) -f1000 -D0 -I100000 -i64 -r 40 -S10 -F2112 -T32 -l0 -o 0

File: library/index.c (hunk 4 above)
Comment (bioothod) [inline library/index.c]: bsearch has to run over single block (4k in size) only. There is a range index tree which contains ranges of keys and their start/stop positions in index. This range tree is generated on eblob start. When we have to find a key, we find a range containing this key and perform binary search in it. Ranges do not overlap, but can contain multiple key entries because of object removal for example.

You should not worry about random read pattern - linux kernel has readahead subsystem which analyzes read pattern and reads ahead part of the file into page cache. Also, index files are frequently accessed and do not flushed out of page cache.

Similar logic should be implemented with pread() approach - read not the whole file but only a chunk containing range, which wraps over they key we are interested in, and then use part of the old eblob_find_on_disk() which ran bsearch(). You can only replace search_start/search_end setting - not from mapped area, but reading it from file descriptor.

Comment (abudnik) [inline library/index.c]: >read not the whole file but only a chunk containing range
Yes, this is what eblob_find_on_disk() actually does in this PR. I made a mistake saying that it reads whole file in my previous comment.

Comment (bioothod) [inline library/index.c]: It does more that that, because you changed the logic within loops, so I asked what it is supposed to do and why did you change that?

File: bindings/python/eblob_python.cpp (hunk 2 above)
Comment (abudnik) [inline bindings/python/eblob_python.cpp]: This binding is not written using newest C++11 standard. Also auto_ptr is convenient and safe (it plays simple role here, without auto_ptr assignments or using it as a function argument) here.

File: bindings/python/eblob_python.cpp (hunk 1 above)
Comment (abudnik) [inline bindings/python/eblob_python.cpp]: > How will it behave when I press Ctrl-C while python binding reads data from fd?

As far as I know ignoring EINTR after syscalls will not affect unix process signal handling. So if SIGINT happens, then process will do some default action (process interrupt) or will do something else (e.g. ignoring signal or something user-defined).

Comment (abudnik) [inline bindings/python/eblob_python.cpp]: >Why did you copy-paste this whole function from one already present in eblob code?

Saying the truth I don't know whether it will be a good idea to move __eblob_read_ll to eblob's public interface. And I am not sure about necessity of this wrapper (eblob_read_ll).
I can use only single pread without any wrappers here.

File: library/index.c (hunk 4 above)
Comment (abudnik) [inline library/index.c]: Logically these loops does same thing as loops in previous implementation without pread.
Loops in this PR reads by blocks items, but can be simplified to reading items one by one. Reading items one by one may be less efficient.
Should I add more comments in this loops or should I rewrite it to less efficient algorithms (reading one by one)?

File: bindings/python/eblob_python.cpp (hunk 1 above)
Comment (bioothod) [inline bindings/python/eblob_python.cpp]: Client code can be interrupted by various reasons and the main one is usually to stop processing.
There are no dedicated signal handlers in client code (compared to server one) except ignoring some signals in workers, which brings me to this question - what will prevent client code from running forever ignoring signals, since it returns back into syscall right after it has ignored signal?

At the server-side there is no such problem, since there is a signal handler which will set exiting flags, and thread which has been interrupted will return to higher level and eventuall will check this flag, and thus will exit.

File: library/index.c (hunk 4 above)
Comment (bioothod) [inline library/index.c]: There are multiple pread() calls in this function, there should be only one right after eblob_index_blocks_search_nolock() like the one at 378 line.

There should be no other reads from index file descriptor like at lines 428 and 472.
Looks like you do not use information obtained from range index tree and reads index data outside of the range potentially covering needed key.

File: bindings/python/eblob_python.cpp (hunk 1 above)
Comment (abudnik) [inline bindings/python/eblob_python.cpp]: >what will prevent client code from running forever ignoring signals, since it returns back into syscall right after it has ignored signal?

Client code may be a server side with specific signal handling.

Comment (bioothod) [inline bindings/python/eblob_python.cpp]: It is not right now, so was the question.

File: library/index.c (hunk 4 above)
Comment (abudnik) [inline library/index.c]: It is possible to check whether interval contains key before index block reading, but additional pread may be needed anyway.

Comment (bioothod) [inline library/index.c]: Why do you need additional read?

After you have read the whole index range (40 index entries by default) into the memory, you run bsearch over this array and optionally linear search if there are duplicates. That is what previous code did.

Comment (abudnik) [inline library/index.c]: end = bctl->sort.data + bctl->sort.size;
start = bctl->sort.data;
...
while (sorted < end && eblob_disk_control_sort(sorted, dc) == 0) {

So it may do linear search through whole index file, not only in single index block. Index file may contain keys with same id.

Comment (bioothod) [inline library/index.c]: No, it doesn't. Please read what I wrote again or check the code exactly 3 lines below what you have shown.

There is a range index tree which splits index file into ranges of 40 keys by default. These ranges are being looked for in eblob_index_blocks_search_nolock() function. It returns pointer to the control structure, which contains start and end keys for range of keys which covers requested key. Code then checks mapped area within those start/end keys and runs bsearch(). It is possible that there are several the same keys in the index file, thus there is linear search right after bsearch().

You broke that logic basically by not taking into account range index tree. I introduced this tree years ago exactly to solve  problem with multiple random reads over the whole index file. Now it is replaced with single jump into the index file and limit its search by the size of the index range (40 entries by default).

Comment (abudnik) [inline library/index.c]: No, I am not mistaken.
search_start != start
search_end != end
And these ^ inequalities stays true before and after bsearch() call.

start, end - range of whole index file
search_start, search_index - range of index block

So my previous comment is correct.

Comment (bioothod) [inline library/index.c]: I made a mistake, bsearch() is used to locate block within range tree (it is sorted array filled at eblob start) if bloom filter didn't reject it first, and then linear search is used to locate given key in the index block (40 entries by default) if it exists.

Comment (bioothod) [inline library/index.c]: Please, read it again. There is no linear search over the whole index. At all. Linear search is performed over 2-3 entries at most.

I'm not going to argue more trying to prove this, didn't I show you the code and write explanation?
If you do not want to understand the code you are trying to change, I can not accept such pull requests.
If you have any further questions on how eblob was designed you are welcome, but I will not repeat myself again.

Comment (bioothod) [inline library/index.c]: In the worst possible scenario linear search could run away from within block boundaries (not over the whole file), but it requires that index repeatedly contains only the same key up to the end of the index file. This is not the case.

Your changes at 413+ lines were added to address this issue, weren't they?

Comment (abudnik) [inline library/index.c]: >This is not the case.

If a few records with the same key located both at the end of first index block and at the beginning of subsequent index block, then additional pread required. 

> changes at 413+ lines were added to address this issue, weren't they?

Yes. This lines simply adopts logic that was here in mmap version to pread imlementation.

Comment (abudnik) [inline library/index.c]: It seems that eblob_find_on_disk() code may cause misunderstoods of what it actually does. This code looks somewhat tricky and hardcore. I will think how to refactor it.

Comment (abudnik) [inline library/index.c]: Also I changed semantics of "sorted" and "end" variables - and this caused all those questions above and confusion. Previously "end" pointed at the end of mapped index file and logically this code potentially does full scan of index file. Now "end" pointed to the end of a buffer, but logic stays the same - with scanning until reaching end of file or finding valid index.

Comment (bioothod) [inline library/index.c]: It could be great to have more commends on how and *why* there are particular chunks of code, please add them into the pull request

Comment (abudnik) [inline library/index.c]: Added comments in commit https://github.com/abudnik/eblob/commit/aae3276062e0a4066a049aa509f2b33a1532cfb8

File: bindings/python/eblob_python.cpp (hunk 1 above)
Comment (shaitan) [inline bindings/python/eblob_python.cpp]: With `eblob` there are no clients or servers. Why do we have different logics for C++ and python bindings?

File: library/index.c [hunk 8]
```diff
@@ -352,27 +353,43 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	bytes = pread(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
+	if (bytes != hdr_block_size) {
+		if (bytes < 0)
+			err = -errno;
+		goto err_out_free_index;
+	}
```
Comment (shaitan) [inline library/index.c]: Again we have only one chance to read whole data. Why doesn't it use `__eblob_read_ll` ?

**Comment by shaitan** [comment]
Is this pull request ready to be merged? If not what should be done?

**Comment by abudnik** [comment]
I think it is ready to be merged.
Is there anything to be fixed?

**Comment by shaitan** [comment]
@bioothod can you merge it?

**Comment by bioothod** [comment]
I will review it today or tomorrow and either merge it or let you know if something can be changed

File: library/index.c [hunk 9]
```diff
@@ -380,30 +395,63 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
 			break;
 		}
 		st->additional_reads++;
-		sorted++;
+
+		if (++sorted > end) {
+			/* read next block of headers */
+			hdr_block_offset += hdr_block_size;
+			if (hdr_block_offset >= bctl->sort.size)
+				break;
```
Comment (bioothod) [inline library/index.c]: hdr_block_size is a multiplication of index blocks and header size, not a single header size, why do you break out of the searching loop here - it is possible that there are not X headers at the very end of the index, but only half of that, and yet they will not be processed, right?

File: library/index.c [hunk 10]
```diff
@@ -380,30 +395,63 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
+			num = (bctl->sort.size - hdr_block_offset) / hdr_size;
+
+			if (num > b->cfg.index_block_size)
+				num = b->cfg.index_block_size;
+
+			hdr_block_size = num * hdr_size;
+			sorted = hdr_block;
+			end = sorted + (num - 1);
+
+			read_err = __eblob_read_ll(bctl->sort.fd, hdr_block, hdr_block_size, hdr_block_offset);
```
Comment (bioothod) [inline library/index.c]: Please add error log after **every** invocation of this function and put as much information as possible: index number, total size, idnex size, position, read position, size and so on. As much information as possible.

File: library/index.c [hunk 11]
```diff
@@ -412,10 +460,49 @@ static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
...
+			break;
+		}
+
+		/*
+		 * if it is first and initial block read then start searching record starting
+		 * from record preceded to founded by bsearch record (sorted_orig),
+		 * otherwise starting from last block record
+		 */
+		if (hdr_block_offset == saved_hdr_block_offset) {
+			sorted = sorted_orig - 1;
```
Comment (bioothod) [inline library/index.c]: Can sorted_orig - 1 be an invalid pointer, i.e. sorted_orig points to the very first header?

Comment (abudnik) [inline library/index.c]: >Can sorted_orig - 1 be an invalid pointer, i.e. sorted_orig points to the very first header?

Yes, but guard condition
if (sorted >= hdr_block) { - will check "sorted" validity.

Ok, I will add more comments

File: library/index.c (hunk 9 above)
Comment (abudnik) [inline library/index.c]: >it is possible that there are not X headers at the very end of the index, but only half of that, and yet they will not be processed, right?

No, because hdr_block_offset means already processed bytes. It increments hdr_block_offset, thereby pointing to the beggining of last block (which is only half of initial hdr_block_size).

**Comment by abudnik** [comment]
Fixes: https://github.com/abudnik/eblob/commit/aa0bfbfa9e47d951adc3d61d449cb9f9d27d4526

File: library/index.c (hunk 9 above)
Comment (bioothod) [inline library/index.c]: Should ++sorted > end be >= instead?

Comment (abudnik) [inline library/index.c]: No, because 'end' points at the last element here.
